README.md
# NRYF Dumper - CLI tool to web scrap chileans personal information

![][BADGE-TEST-MAIN] ![][BADGE-BUILD-MAIN] ![][BADGE-BUILD-ART-MAIN] ![][BADGE-PUBLISH-MAIN]

## Disclaimer
NRYF Dumper is not affiliated by any means to the following information sources and it's creators: "elrutificador.com". The purpose of the code within this repository is only to query the already mentioned information sources without the need of a browser.

The author have no responsibility at all by the derived usage of this respository's code.

## Install
### Via npm
```
npm install -g nombrerutyfirma-dumper
```
### Using source code
```
git clone https://github.com/iolave/nombrerutyfirma-dumper.git
cd nombrerutyfirma-dumper
npm ci
npm run build
npm install -g .
```

## Usage
```
nryf-dumper --source=[source] --output=[output] [output options] [global options] [input options]
```

**Examples**
```shell
# ########################################
# Query a single rut and output to console
# ########################################
nryf-dumper --source=elrutificador --rut=123456 --output=console
# 2023-08-16T16:04:25.242Z [INFO]    elrutificador: data not found for rut 123.456-0

# ########################################
# Query multipe ruts and output to a file
# ########################################
nryf-dumper --source=elrutificador --ruts=123456,1234567 --output=local-file  --out-path=/tmp/results.json
# 2023-08-16T16:06:42.249Z [INFO]    elrutificador: data not found for rut 123.456-0, skipping
# 2023-08-16T16:06:44.252Z [INFO]    elrutificador: elrutificador: wrote found data for rut 1.234.567-4 to /tmp/results.json
```

### Available sources
- elrutificador.com: `elrutificador`
- nombrerutyfirma.com: `nombrerutyfirma`
- rutificador.net: `rutificador-net`

### Available outputs
- Console: `console`
- Local file: `local-file`
    - Required option: `--out-path=[path] # i.e. --out-path=/tmp/file.json`
- MongoDB: `mongodb`
    - Required option: `--uri=[connection string] # i.e. --uri=mongodb://127.0.0.1/`


### Available input options
- Single rut: `--rut=[rut without dv] # i.e. --rut=123456`
- Range of ruts: `--from-rut=[rut without dv] --to-rut=[rut without dv] # i.e. --from-rut=123450 --to-rut=123460`
- Multiple ruts: `--ruts=[...ruts without dv] # i.e. --ruts=123456,123457,123458`

### Global options
- Verbosity: `--verbose | --verbose=[number]`
- Batch size (for parallel requests): `--batch-size=[number] # i.e. --batch-size=10`
- Max retries: `--max-retries=<number> # i.e. --max-retries=0`

[BADGE-TEST-MAIN]: https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/wiki/iolave/nombrerutyfirma-dumper/test-main.md&logo=github
[BADGE-BUILD-MAIN]: https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/wiki/iolave/nombrerutyfirma-dumper/build-main.md&logo=github
[BADGE-BUILD-ART-MAIN]: https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/wiki/iolave/nombrerutyfirma-dumper/build-artifacts-main.md&logo=github
[BADGE-PUBLISH-MAIN]: https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/wiki/iolave/nombrerutyfirma-dumper/publish-main.md&logo=github

## Logging
NRYF Dumper logging is currently available for MacOS and Linux.

These are the log paths for each OS:
- MacOS: `$HOME/Library/Logs/nryf-dumper.log`
- Linux: `/var/log/nryf-dumper.log`


CHANGELOG.md
# nombrerutyfirma-dumper

## 1.4.1

### Patch Changes

- Fixed the `--batch-size` option on `--output=mongodb` so now it's working

## 1.4.0

### Minor Changes

- 11fdf4d: Added commander mongodb action
- 641f79c: Added logging for MacOS and Linux
- 845dec7: Added `--option=mongodb` and `--uri=[connection-string]` options
- 593df69: Added `rutificador-net` information source

### Patch Changes

- 27fd607: Fixed a bug on local-file:single-rut `nombrerutyfirma` source that used `elrutificador` source instead of the proper one
- 536f05d: Fixed a bug that caused the local-file writestream to close on a loop cycle

## 1.3.0

### Minor Changes

- b44eb10: Added `nombrerutyfirma.com` search by rut method
- 6877f81: Added `nombrerutyfirma` cli source

### Patch Changes

- 4e42eb2: Added `node-fetch-commonjs` dependency as it includes `FetchError`
- 66a23e3: Added scoped NRYFError error codes to have more control over errors
- 761ffa0: Renamed `elrutifcador` source internal directory to `el-rutificador`

## 1.2.1

### Patch Changes

- 59d6e0f: Bumped `@types/node` dev dependency version to `20.5.9`

## 1.2.0

### Minor Changes

- Integrated elrutificador table generation v2
- Added retry on banned_ip response from cloudflare
- Integrated new elrutificador jwt generation v2
- Added an error class that will allow to standarize errors across nryf-dumper

### Patch Changes

- Bumped `typescript` dev dependency version to `5.2.2`
- Bumped `@types/node` dev dependency version to `20.5.7`

## 1.1.0

### Minor Changes

- Added `--max-retries <number>` option that will retry on failure for some specific scenarios
- Added `elrutificadorByRut()` retry handler

### Patch Changes

- Added `package.json` keywords.
- Added `version:changeset` and `version:bump` npm scripts

## 1.0.3

### Patch Changes

- Bumped `typescript` dev dependency version to 5.1.6

## 1.0.2

### Patch Changes

- Bumped `@types/node` dev dependency version to 20.5.3

## 1.0.1

### Patch Changes

- Bumped typescript patch version.
- Replaced `eg:` to `i.e.:` within options descriptions.
- Modified the `--verbose` option description.

## 1.0.0

### Major Changes

- Added `--source` flag with support for `elrutificador.com` information source.
- Added `--output` flag with support for `console` and `local-file`.
- Added `--rut` flag to query a single rut.
- Added `--ruts` flag to query multiple ruts.
- Added `--from-rut` and `--to-rut` flags to query a range of ruts.
- Added `--verbose` flag to output more information through the console.
- Added `--batch-size` flag to allow simultaneous requests.
- Published package to npm.


src/cli/index.ts
import { Option, program } from "commander";
import { intParser, numberArrayParser } from "./parser";
import { InformationSource, Destination, destinations, informationSources } from "../business-logic/commander";
import { setLogLevel } from "../config/logger";
import packageJson from "../../package.json"

export default function parseProgram(): ProgramOptions {
    program.parse();

    const args = program.opts<ProgramOptions>();

    /* LOGGER LEVER */
    if (!args.verbose) setLogLevel(6); // 6 refers to info
    else if (typeof args.verbose === "boolean") args.verbose ? setLogLevel(7):setLogLevel(6); // 7 refers to debug
    else setLogLevel(args.verbose); // 7 refers to debug

    if (args.output === "console") {
        // CONFLICTS
        if (args.outPath) program.error(`error: option '${outPathOptionFlags}' can not be used alongside '--output ${args.output}'`)
        if (args.uri) program.error(`error: option '${uriOptionFlags}' can not be used alongside '--output ${args.output}'`)
    }
    else if (args.output === "local-file") {
        // CONFLICTS
        if (args.uri) program.error(`error: option '${uriOptionFlags}' can not be used alongside '--output ${args.output}'`)
        // REQUIRED
        if (!args.outPath) program.error(`error: required option '${outPathOptionFlags}' not specified`)
    }
    else if (args.output === 'mongodb') {
        // CONFLICTS
        if (args.outPath) program.error(`error: option '${outPathOptionFlags}' can not be used alongside '--output ${args.output}'`)
        // REQUIRED
        if (!args.uri) program.error(`error: required option '${uriOptionFlags}' not specified`)
    }

    if (args.queryType === "single-rut") {
        if (!args.rut) program.error(`error: required option '${rutOptionFlags}' not specified`);
    }
    else if (args.queryType === "ruts-range" ) {
        if (!args.fromRut) program.error(`error: required option '${fromRutOptionFlags}' not specified`);
        if (!args.toRut) program.error(`error: required option ${toRutOptionFlags}' not specified`);
    }
    else if (args.queryType === "multiple-ruts") {
        if (!args.ruts) program.error(`error: required option '${rutsOptionFlags}' not specified`);
    }
    else {
        const errorMsg = new String()
            .concat("error: no input specified\n\n")
            .concat("specify one of the following options\n")
            .concat(`  ${rutOptionFlags}\n`)
            .concat(`  ${fromRutOptionFlags}\n`)
            .concat(`  ${toRutOptionFlags}\n`)
            .concat(`  ${rutsOptionFlags}\n`)
        ;

        program.error(errorMsg);
    }

    return args;
}

// TODO: use zod for this
// TODO: or use custom types for each dest
interface ProgramOptionsBase {
    verbose?: boolean | number;
    source: InformationSource;
    output: Destination;
    outPath?: string;
    uri?: string;
    maxRetries?: number;
}

interface SingleRutProgramOptions extends ProgramOptionsBase {
    queryType: "single-rut",
    rut: number;
}

interface RutsRangeProgramOptions extends ProgramOptionsBase {
    queryType: "ruts-range",
    fromRut: number;
    toRut: number;
    batchSize?: number;
}

interface MultipleRutsProgramOptions extends ProgramOptionsBase {
    queryType: "multiple-ruts",
    ruts: number[];
    batchSize?: number;
}

export type ProgramOptions = SingleRutProgramOptions | RutsRangeProgramOptions | MultipleRutsProgramOptions;

program.name("nryf-dumper");
program.showHelpAfterError();
program.version(packageJson.version);

const verboseOptionFlags = "--verbose [level]";
const verboseOption = new Option(verboseOptionFlags, "set verbosity level (see RFC5424)");
verboseOption.argParser<number>((value: string) => intParser(program, value, verboseOptionFlags));

const sourceOptionFlags = "--source <source>";
const sourceOption = new Option(sourceOptionFlags, "information source");
sourceOption.choices(informationSources);
sourceOption.makeOptionMandatory(true);

const outputOptionFlags = "--output <dest>";
const outputOption = new Option(outputOptionFlags, "write to the desired output");
outputOption.choices(destinations);
outputOption.makeOptionMandatory(true);

const outPathOptionFlags = "--out-path <path>";
const outPathOption = new Option(outPathOptionFlags, "out path");
outPathOption.makeOptionMandatory(false);

const uriOptionFlags = "--uri <connection string>";
const uriOption = new Option(uriOptionFlags, "database connection string (destination: [mongodb])");
uriOption.makeOptionMandatory(false);

const rutOptionFlags = "--rut <rut without dv>";
const rutOption = new Option(rutOptionFlags, "search a single rut");
rutOption.argParser((value) => intParser(program, value, rutOptionFlags));
rutOption.makeOptionMandatory(false);
rutOption.conflicts("fromRut");
rutOption.conflicts("toRut");
rutOption.conflicts("ruts");
rutOption.implies({ queryType: "single-rut" });

const fromRutOptionFlags = "--from-rut <rut without dv>";
const fromRutOption = new Option(fromRutOptionFlags, "search a bunch of ruts starting from this rut. i.e.: 1");
fromRutOption.makeOptionMandatory(false);
fromRutOption.conflicts("rut");
fromRutOption.conflicts("ruts");
fromRutOption.argParser((value) => intParser(program, value, fromRutOptionFlags));
fromRutOption.implies({ queryType: "ruts-range" });

const toRutOptionFlags = "--to-rut <rut without dv>";
const toRutOption = new Option(toRutOptionFlags, "search a bunch of ruts ending with this rut. i.e.: 1000000");
toRutOption.makeOptionMandatory(false);
toRutOption.conflicts("rut");
toRutOption.conflicts("ruts");
toRutOption.argParser((value) => intParser(program, value, toRutOptionFlags));
toRutOption.implies({ queryType: "ruts-range" });

const rutsOptionFlags = "--ruts <...ruts without dv>";
const rutsOption = new Option(rutsOptionFlags, "search a bunch of ruts. i.e.: 123456,123457,1234568");
rutsOption.makeOptionMandatory(false);
rutsOption.conflicts("rut");
rutsOption.conflicts("fromRut");
rutsOption.conflicts("toRut");
rutsOption.argParser((value) => numberArrayParser(program, value, rutsOptionFlags));
rutsOption.implies({ queryType: "multiple-ruts" });

const batchSizeOptionFlags = "--batch-size <number>";
const batchSizeOption = new Option(batchSizeOptionFlags, "number of simultaneous requests");
batchSizeOption.makeOptionMandatory(false);
batchSizeOption.conflicts("rut");
batchSizeOption.argParser((value) => intParser(program, value, batchSizeOptionFlags));

const maxRetriesOptionFlags = "--max-retries <number>";
const maxRetriesOption = new Option(maxRetriesOptionFlags, "number of max retries on specific scenarios. defaults to infinte");
maxRetriesOption.makeOptionMandatory(false);
maxRetriesOption.argParser((value) => intParser(program, value, maxRetriesOptionFlags));

program.addOption(verboseOption);
program.addOption(sourceOption);
program.addOption(outputOption);
program.addOption(outPathOption);
program.addOption(uriOption);
program.addOption(rutOption);
program.addOption(fromRutOption);
program.addOption(toRutOption);
program.addOption(rutsOption);
program.addOption(batchSizeOption);
program.addOption(maxRetriesOption);
